### ISSUE TITLE
Add configuration option for diagnostic delays in editor

### ISSUE THREAD

**Issue by cartermp** [dotnet/roslyn#19347]
This is from this comment: https://github.com/dotnet/roslyn/issues/18376#issuecomment-299944799

F# is hurt a bit by artificial delays in squiggles in the editor.  Because we have type inference, people rely on instant feedback on if something will work via squiggles (or lack thereof).  Because type inference already takes quite a bit of time, any additional delay gives the appearance the editor is unresponsive in certain scenarios for F#.

A configuration flag to opt out of these delays would be great.  We would configure this in our editor layer.

**Comment by isaacabraham** [comment]
Perhaps a slightly nicer option would be to be able to set the _minimum_ delay. @CyrusNajmabadi makes a valid point that I've noticed in the past - if you get *instant* feedback whilst typing and get squigglies in the middle of typing, it can be annoying. If it was possible to set the delay at no less than e.g. 150ms this would avoid that concern, and also (hopefully) help ensure that the IDE responsiveness was more or less consistent across different files.

**Comment by Pilchie** [comment]
Tagging @heejaechang and @jinujoseph about this request.

**Comment by heejaechang** [comment]
it is a bit hard since that delay is shared by every one. not just F#.

**Comment by CyrusNajmabadi** [comment]
We can certainly make it configurable on a per-feature basis.

**Comment by isaacabraham** [comment]
@heejaechang optional argument maybe?

**Comment by heejaechang** [comment]
presentation can be slowed. but actual work (computing diagnostics) can't be controlled per language or feature right now. the delay for actual work is there to reduce work we do on each key stroke. otherwise, we will produce too much garbage that will affect typing responsiveness. (allocation rate goes up, GC runs more often and etc)

current delay is set to current number after years of tweaking from our typing responsiveness perf result.

but that being said, we could (even though it is some work) make delay to be per language somehow since each language might have different cost on computing diagnostics. for C# and VB, that is most expansive thing to compute.

**Comment by dsyme** [comment]
@heejaechang @CyrusNajmabadi The current delay for F# "document analysis" in order to update red squigglies is really, really long.   We have quite a problem here

The frustrating thing is that other analysis such as semantic happens very quickly - and for F# the semantic colorization involves doing all the work necessary for document analysis (red squigglies).  So F# is all ready to report the errors, but Roslyn takes a long time before the errors are shown

I would really like to make progress on this.  The default settings that Roslyn are using appear to be wrong for F# - people using F# expect much faster red-squiggly appearance than we're currently getting.  Perhaps it is our fault (are the previous stages taking too long?), but I feel it is also to do with Roislyn (e.g. is Roslyn analyzing the wrong files, rather than concentrating on the active buffers???)

**Comment by heejaechang** [comment]
@dsyme can you experiment with numbers to see what affects your scenario. if you build Roslyn.sln, and run devenv /rootsuffix Roslyn you will see this in options.

![image](https://user-images.githubusercontent.com/1333179/27355894-62f16424-55c1-11e7-82ea-180de20fcd27.png)

try reduce Active File worker backoff timespan in ms to something like 100ms or lower. you probably want some delay there still to reduce allocations.

we arrived to 400ms to reduce typing response time due to garbage produced by diagnostics.

**Comment by Pilchie** [comment]
Updated comment above to say `RoslynDev` instead of `Roslyn` for the rootsuffix argument.

**Comment by dsyme** [comment]
@heejaechang I'll sarch the code later (not online right now) but do you have any links for how Roslyn determines the "active file"? I'm wondering if that could be contributing if it's not always being set correctly by the requests F# makes

**Comment by heejaechang** [comment]
@dsyme we get from here.

http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/Workspace/VisualStudioDocumentTrackingService.cs,42

but, this is general Roslyn workspace service (VS layer),so I dont think it will affect certain language differently.

**Comment by dsyme** [comment]
@heejaechang Thanks, that gives me a fair bit to go on

Note that if you look at https://github.com/Microsoft/visualfsharp/issues/3094#issuecomment-310163887 we are seeing 8-10second delays before semantic colorization kicks in after a change in the solution configuration.  This is just one example of long delays.

Is there any way to trace what Roslyn thinks is going on (i.e. what tasks it is scheduling and when "back offs" are applied). Or is there one particular routine where you would recommend we put some tracing output?

Thanks

**Comment by heejaechang** [comment]
is the issue semantic colorizer or syntax colorizer?

see this for when semantic colorizer will re-colorize the view
http://source.roslyn.io/#Microsoft.CodeAnalysis.EditorFeatures/Implementation/Classification/SemanticClassificationViewTaggerProvider.cs,55

for syntax colorizer
http://source.roslyn.io/#Microsoft.CodeAnalysis.EditorFeatures/Implementation/Classification/SyntacticClassificationTaggerProvider.TagComputer.cs,407

you can set BP on the links above to see who asks to recolorize view and why.

**Comment by dsyme** [comment]
@heejaechang semantic.  Thanks for the links, it's super helpful :)

**Comment by dsyme** [comment]
@heejaechang Does Roslyn have a work queue which for some reason gets "clogged up" with requests/operations?  If so, is there a place I could put tracing about the number of items in the queue, their priority and so on?

Another possibility is this... We currently call [analyzerService.Reanalyze](https://github.com/Microsoft/visualfsharp/blob/master/vsintegration/src/FSharp.Editor/LanguageService/LanguageService.fs#L62) when the F# Compiler Service has detected a change to its semantic conditions for a file (we don't rely on Roslyn's "semantic" change events, but instead we use this technique).  

Will these analyzerService.Reanalyze requests be prioritized, or will they get the very lowest priority?  If the latter, is there any way we can request a reanalysis at high priority?

thx again

**Comment by heejaechang** [comment]
@dsyme what is actual type of that analyzer service? is that this? http://source.roslyn.io/#Microsoft.CodeAnalysis.Features/Diagnostics/IDiagnosticAnalyzerService.cs,16

if so, you can give in flag saying it is high priority. (* NOTE *, if there are too many people claiming theirs are high priority, we will simply drop those requests and put it as normal, so dont make everything you guys own as higher priority)

... 

about queue.

http://source.roslyn.io/#Microsoft.CodeAnalysis.Features/SolutionCrawler/WorkCoordinator.IncrementalAnalyzerProcessor.cs,88c5ecec53239e9f

this is entry point.

**Comment by dsyme** [comment]
> @dsyme what is actual type of that analyzer service? is that this? http://source.roslyn.io/#Microsoft.CodeAnalysis.Features/Diagnostics/IDiagnosticAnalyzerService.cs,16


@heejaechang Yes that's it.  Nice to see the flag is so simple!

One problem is that we only want it "highPriority" if the document is the current active view document in VS.  But we can work out how to work that out (unless there is some simple Roslyn way to do that)

**Comment by heejaechang** [comment]
@dsyme I believe you guys are using DocumentDiagnosticAnalyzer? and has IVT to use? then use this.
http://source.roslyn.io/#Microsoft.CodeAnalysis.Features/SolutionCrawler/IDocumentTrackingService.cs,15

**Comment by heejaechang** [comment]
@dsyme without actually debugging through your code, I can't say for sure, but if a document is the one that has a focus, that document automatically becomes high priority document... so.. there seems to be something I am missing

**Comment by dsyme** [comment]
> @dsyme without actually debugging through your code, I can't say for sure, but if a document is the one that has a focus, that document automatically becomes high priority document... so.. there seems to be something I am missing

OK, thanks.  Yes, setting highPriority doesn't seem to change anything

**Comment by dsyme** [comment]
@heejaechang Do we get any control over the priority order for analyzers?

What I'm seeing is the our analyzers are called in this order after a Reanalyze request:

    devenv.exe Information: 0 : SimplifyName
    devenv.exe Information: 0 : UnusedDeclarationsAnalyzer
    devenv.exe Information: 0 : UnusedOpens
    devenv.exe Information: 0 : SemanticColorization

Ideally we want SemanticColorization to have higher priority

Also, I note these aren't being called in parallel after a Reanalyze request (which seems to be what happens after a user edit).  

Also, I see "HighPriorityForActiveFile" here - could there be a chance we are not activating that?  I don't see if that is being set for DefaultDiagnosticAnalyzerService, e.g. http://source.roslyn.io/#microsoft.codeanalysis.workspaces/SolutionCrawler/ExportIncrementalAnalyzerProviderAttribute.cs,34

**Comment by dsyme** [comment]
@heejaechang For example. we declare the SimplifyName analyzer like this:

```
[<DiagnosticAnalyzer(FSharpConstants.FSharpLanguageName)>]
type internal SimplifyNameDiagnosticAnalyzer() =
    inherit DocumentDiagnosticAnalyzer()
```

We want this one to be low priority

**Comment by dsyme** [comment]
@heejaechang Specifically, can some DocumentDiagnosticAnalyzer's be lower priority than others?  Or can we make the ones we want to be lower priority into some other kind of analyzer?

**Comment by heejaechang** [comment]
@dsyme we never ran analyzers concurrently in VS. we do in OOP, but not inproc. due to allocations.

right now, we don't support analyzer to specific priority, but I guess we can provide one for Document/ProjectDiagnosticAnalyzer since they are internal only.

**Comment by dsyme** [comment]
> @dsyme we never ran analyzers concurrently in VS. we do in OOP, but not inproc. due to allocations.

Yes, we (ok, well, I) misunderstood what was happening here.  Our working assumption had been that analyzers were being triggered in parallel and, in particular, we could have some relatively slow DocumentDiagnosticAnalyzers without hurting the main ones which present reg-squigglies.

**Comment by cartermp** [comment]
Just to provide an example from an issue tracking this on our side, this delay really degrades the experience for F#:

![](https://user-images.githubusercontent.com/873919/27695441-8f8d6112-5cf7-11e7-82fd-2df33162d3aa.gif)

The above gif is on a fast machine as well.  I'm personally seeing better times on a fresh install, but I can still see that there's a delay here when it's not necessary.

**Comment by CyrusNajmabadi** [comment]
@heejaechang Do we know why this is happening.  This is not an issue, for example, for TypeScript, despite them using the same interfaces that F# is using.

**Comment by heejaechang** [comment]
@CyrusNajmabadi @dsyme already did the analysis and wrote why it is happening - https://github.com/dotnet/roslyn/issues/19347#issuecomment-310225540

also more detail in here - https://github.com/dotnet/roslyn/issues/20390

@dsyme did you open a bug to support priority between document/project diagnostic analyzers?

**Comment by dsyme** [comment]
> @dsyme did you open a bug to support priority between document/project diagnostic analyzers?

@heejaechang  This is an old thread - but no, I didn't open a bug on this.  I see you've covered it in https://github.com/dotnet/roslyn/issues/19347 to some extent.  This option would definitely be useful for F#, as would running the analyzers in parallel in the "reanalyze" situation described above

**Comment by vasily-kirichenko** [comment]
@heejaechang I cannot see "Roslyn" node in settings in VS instance running with `/rootsuffix RoslynDev /log` arguments. Has this feature been removed?

**Comment by heejaechang** [comment]
@vasily-kirichenko i should be still there. not sure what you are trying to do with it though.

**Comment by isaacabraham** [comment]
What was the outcome here - did this ever get resolved?

**Comment by cartermp** [comment]
The core issue was that diagnostic analysis was not guaranteed to kick off and complete. This was because:

1. Document analyzers are in-process, like F#, and they are run serially. Running them in parallel has resulted in significant performance degradation of the entire VS IDE
2. The Unused Opens analyzer had a performance issue.
3. There was no priority in which to run analyzers. This lack of priority meant that there was no guarantee when diagnostic analysis would be run, which meant that you could get feedback on a compile error anywhere from 2 seconds to 10+ seconds if a more expensive analyzer (e.g., Unused Opens) was ran first.

This has since been fixed in VS 15.7 by running these in-process analyzers in a priority order that F# defines. The Unused Opens analyzer was also improved.

In terms of day-to-day use of the VF# codebase, editor feedback is in a good place, and I believe that the root of our remaining problems lies in our own codebase. There is still a delay toggle that could be added, so this issue is **not** resolved.

**Comment by heejaechang** [comment]
we could redesign and provide a way for each language, analyzers to have different delay. but that will be a bit more work than bug level. 

we can keep this bug in case we want to do that work.

tagging @jinujoseph

**Comment by kendrahavens** [comment]
This user would like to increase the delay when receiving diagnostics so making this configurable would definitely help their use case: [Developer community: Add Delay in Real Time Compilation in Visual Studio](https://developercommunity.visualstudio.com/content/idea/417970/add-delay-in-real-time-compilation-in-visual-studi.html )

**Comment by CyrusNajmabadi** [comment]
I'm not certain i agree.  User says:

>I am working on an Open Source Project based on C# using Visual Studio 2017. Whenever I make any changes in it, Visual Studio (in a delay of about half a second) starts checking the code for errors in it which is why sometimes it gets hanged

There shoudl be no reason that VS gets "hung" due to errors being computed.  Even if we implement this fix, it is unlikely to help the user out for this scenario.  We should actually communicate with the user to get real traces to find out waht the real problem is that they're experiencing.

**Comment by kendrahavens** [comment]
@CyrusNajmabadi I agree with that, I've requested a trace in the devcomm feedback. I was focusing more on the irritation coming from users that do not have a fast typing speed.

**Comment by CyrusNajmabadi** [comment]
@kendrahavens makes sense to me :)

**Comment by CyrusNajmabadi** [comment]
Roslyn does not control the delay here.  It is set by VS as the LSP host.  Roslyn just returns the diagnsotic data whne queried.  If people want this to be configurable, it needs to be a request made through developer community that the VS team can decide on.  Closing as external.
